dipy/reconst/ivim.py

```diff
@@ -0,0 +1,285 @@
+                                       algorithm, gtol, ftol, eps)
+        elif fit_method == "two_stage":
+            params_in_mask = two_stage(data_in_mask, self.gtab, x0,
+                                       self.split_b, jac, bounds, tol,
+                                       routine,
+                                       algorithm, gtol, ftol, eps)
+        else:
+            raise ValueError("""Fit method must be either
+                             'one_stage' or 'two_stage'""")
+
+        if mask is None:
+            out_shape = data.shape[:-1] + (-1, )
+            ivim_params = params_in_mask.reshape(out_shape)
+        else:
+            ivim_params = np.zeros(data.shape[:-1] + (4,))
+            ivim_params[mask, :] = params_in_mask
+
+        return IvimFit(self, ivim_params)
+
+
+class IvimFit(object):
+
+    def __init__(self, model, model_params):
+        """ Initialize a IvimFit class instance.
+            Parameters
+            ----------
+            The model parameters are S0, f, D_star, D
+        """
+        self.model = model
+        self.model_params = model_params
+
+    def __getitem__(self, index):
+        model_params = self.model_params
+        N = model_params.ndim
+        if type(index) is not tuple:
+            index = (index,)
+        elif len(index) >= model_params.ndim:
+            raise IndexError("IndexError: invalid index")
+        index = index + (slice(None),) * (N - len(index))
+        return type(self)(self.model, model_params[index])
+
+    @property
+    def shape(self):
+        return self.model_params.shape[:-1]
+
+
+def one_stage(data, gtab, x0, jac, bounds, tol, routine, algorithm,
+              gtol, ftol, eps):
+    """
+    Fit the ivim params using minimize
+
+    Parameters
+    ----------
+    data : array ([X, Y, Z, ...], g)
+        Data or response variables holding the data. Note that the last
+        dimension should contain the data. It makes no copies of data.
+
+    jac : bool
+        Use the Jacobian? Default: False
+
+    Returns
+    -------
+    ivim_params: the S0, f, D_star, D value for each voxel.
+
+    """
+    flat_data = data.reshape((-1, data.shape[-1]))
+    flat_x0 = x0.reshape(-1, x0.shape[-1])
+    # Flatten for the iteration over voxels:
+    bvals = gtab.bvals
+    ivim_params = np.empty((flat_data.shape[0], 4))
+
+    if routine == 'minimize':
+        result = _minimize(flat_data, bvals, flat_x0, ivim_params,
+                           bounds, tol, jac, algorithm, gtol, ftol, eps)
+    elif routine == "leastsq":
+        result = _leastsq(flat_data, bvals, flat_x0, ivim_params)
+    ivim_params.shape = data.shape[:-1] + (4,)
+    return ivim_params
+
+
+def _minimize(flat_data, bvals, flat_x0, ivim_params,
+              bounds, tol, jac, algorithm, gtol, ftol, eps):
+    """Use minimize for finding ivim_params"""
+    sum_sq = lambda params, bvals, signal: np.sum(
+        _ivim_error(params, bvals, signal))
```

Comment:
Set `ivim_error` to calculate the difference, and then square it and sum it here.

Then, you can reuse the same function in `leastsq` below